Add CogniteTimeSeries Search Tab

muradsater opened the pull request:
## Overview
Adds a new 'CogniteTimeSeries' tab that uses DMS (Data Modeling Service) search endpoints to find timeseries.

## Key Features
- **DMS Integration**: Uses Data Modeling Service endpoints for timeseries discovery
- **Smart Search**: Real-time search with AsyncSelect pattern
- **Space & View Selection**: Configurable space and view dropdowns
- **Instance ID Support**: Uses instanceId format (space + externalId) for queries
- **Default Configuration**: Pre-configured with cdf_cdm/v1/CogniteTimeSeries

## Technical Implementation
- Positioned as 2nd tab after 'Time series search'
- Uses /models/spaces, /models/views, and /models/instances/search endpoints
- Includes error handling and loading states
- Filters out string-type timeseries automatically
- Compatible with existing visualization features

## Testing
- ✅ Build successful (version 4.4.0)
- ✅ TypeScript compilation passes
- ✅ Development server runs successfully

## Commits
- feat: add DMS API support
- feat: add CogniteTimeSeries search tab type definitions  
- feat: add CogniteTimeSeries search UI component
- feat: integrate with query system
- chore: bump version to 4.4.0

## Screenshot
![image](https://github.com/user-attachments/assets/685ed3f2-3cea-467f-92e1-160546d09d4b)

muradsater committed:
feat: add DMS (Data Modeling Service) API support

- Add DMSSpace, DMSView, DMSInstance, and DMSSearchRequest types
- Add fetchDMSSpaces() API function for listing available spaces
- Add fetchDMSViews() API function for listing views in a space
- Add searchDMSInstances() API function for searching timeseries instances
- Support includeGlobal parameter to access global spaces like cdf_cdm

muradsater committed:
feat: add CogniteTimeSeries search tab type definitions

- Add CogniteTimeSeriesSearch tab enum value
- Add CogniteTimeSeriesSearchQuery interface with space, version, externalId, and selectedTimeseries fields
- Add default query configuration with cdf_cdm/v1/CogniteTimeSeries defaults
- Add instanceId support to DataQueryRequestItem for DMS-based queries
- Add tab title mapping for the new tab

muradsater committed:
feat: add CogniteTimeSeries search UI component

- Create CogniteTimeSeriesSearchTab with space, version, and view dropdowns
- Implement AsyncSelect pattern for combined search and selection
- Add real-time DMS search with filter to exclude string timeseries
- Support custom labels and latest value toggle
- Include error handling and loading states
- Use React hooks for state management and async operations

muradsater committed:
feat: integrate CogniteTimeSeries search with query system

- Add CogniteTimeSeries tab to queryEditor with proper routing
- Integrate tab with main datasource groupTargets and filtering logic
- Add backend support in TimeseriesDatasource for instanceId format
- Support DMS search results conversion to instanceId for datapoint queries
- Add label handling for DMS-based timeseries with selectedTimeseries fallback
- Position tab as 2nd tab after 'Time series search'

muradsater committed:
chore: bump version to 4.4.0

- Update package.json version for new CogniteTimeSeries search feature release

muradsater committed:
test: add comprehensive unit tests for CogniteTimeSeries search functionality

- Add tests for type definitions and default query validation
- Add tests for query validation with custom configurations
- Add tests for instance ID format generation
- Add tests for DMS search request structure
- Add tests for tab configuration and label generation
- Ensure 100% test coverage for new CogniteTimeSeries features

muradsater committed:
fix: add missing newline at end of test file

- Fix ESLint eol-last rule violation

polomani commented on src/__tests__/cogniteTimeSeriesSearch.test.ts:
```diff
@@ -0,0 +1,226 @@
+      };
+
+      // Should fallback to space:externalId format
+      const fallbackLabel = `${query.selectedTimeseries?.space}:${query.selectedTimeseries?.externalId}`;
```
tests in this file don't actually test anything.
best would be to have both an e2e (playwright) and a unit test with mocked http calls.
we have examples of both of these in the repo

muradsater committed:
test: implement proper unit tests with mocked HTTP calls for CogniteTimeSeriesSearch

- Replace static structure tests with functional tests
- Add proper unit tests for fetchDMSSpaces, fetchDMSViews, searchDMSInstances
- Test both success and error scenarios with mocked HTTP calls
- Add integration tests for complete DMS API workflow
- Test empty responses and error handling
- Follow existing test patterns from login.test.ts and other test files
- Address feedback from PR review about testing actual functionality

muradsater committed:
feat: Add comprehensive e2e tests for CogniteTimeSeries functionality

- Fix failing CogniteTimeSeries test in datasource.spec.ts by adding viewport resize and scroll
- Add three new CogniteTimeSeries tests to queryEditor.spec.ts:
  - Tab selection and UI element visibility test
  - Single query with search and selection test
  - Multiple queries test with legend verification
- Fix all selector issues to match actual HTML structure
- Use correct test data (59.9139-based timeseries) for search functionality
- All 15 e2e tests now pass successfully

muradsater committed:
Update weather-station.json dashboard configuration

polomani commented on src/cdf/client.ts:
```diff
@@ -118,6 +124,17 @@ export async function getLabelsForTarget(
+      // For CogniteTimeSeriesSearch, we use the name from the selected timeseries
+      // since instanceId format doesn't work with regular timeseries endpoints
+      if (target.cogniteTimeSeriesSearchQuery?.selectedTimeseries?.name) {
+        const selectedName = target.cogniteTimeSeriesSearchQuery.selectedTimeseries.name;
+        // If we have a custom label with variables, we can't inject properties from regular timeseries metadata
+        // since this is a DMS instance, so we'll use the selected name as fallback
+        return labelSrc && !labelContainsVariableProps(labelSrc) ? [labelSrc] : [selectedName];
```
we need to call instance by id. we should not save the name in the JSON, it can change

polomani commented on src/cdf/client.ts:
```diff
@@ -400,3 +417,57 @@ export function fetchRelationships(
+  });
+}
+
+export function createInstanceDataQueryRequest(
+  instances: DMSInstanceId[],
+  options: QueryOptions & { timeZone: string }
+): DataQueryRequestItem[] {
+  return instances.map(({ space, externalId }) => ({
+    instanceId: {
+      space,
+      externalId,
+    },
+  }));
+}
```
this function seems redundant

polomani commented on src/cdf/types.ts:
```diff
@@ -263,3 +263,82 @@ export interface RelationshipsFilter {
   limit?: number;
 }
+
+// DMS API types
+export interface DMSSpace {
+  space: string;
+  name?: string;
+  description?: string;
+  createdTime: number;
+  lastUpdatedTime: number;
+}
```
can be in a separate file maybe? `/types/dms.ts`

polomani commented on src/cdf/types.ts:
```diff
@@ -263,3 +263,82 @@ export interface RelationshipsFilter {
+  };
+  range?: {
+    property: string[];
+    gte?: any;
+    gt?: any;
+    lte?: any;
+    lt?: any;
```
preferrably not any

polomani commented on src/components/cogniteTimeSeriesSearchTab.tsx:
```diff
@@ -0,0 +1,279 @@
+import { CommonEditors, LabelEditor } from './commonEditors';
+
+interface CogniteTimeSeriesSearchTabProps extends SelectedProps {
+  connector: any;
```
not any

polomani commented on src/components/cogniteTimeSeriesSearchTab.tsx:
```diff
@@ -0,0 +1,279 @@
+  connector: any;
+}
+
+const LatestValueCheckbox = (props: SelectedProps) => {
+  const { query, onQueryChange } = props;
+  return (
+    <InlineFieldRow>
+      <InlineField
+        label="Latest value"
+        labelWidth={14}
+        tooltip="Fetch the latest data point in the provided time range"
+      >
+        <InlineSwitch
+          label='Latest value'
+          id={`latest-value-${query.refId}`}
+          value={query.latestValue}
+          onChange={({ currentTarget }) => onQueryChange({ latestValue: currentTarget.checked })}
+        />
+      </InlineField>
+    </InlineFieldRow>
+  );
+};
```
we already have this component, this is a copy-paste

muradsater committed:
Move DMS types to separate file

- Move DMS types (DMSSpace, DMSView, DMSInstance) to /types/dms.ts
- Remove duplicate type definitions from multiple files
- Improve code organization per PR feedback

muradsater committed:
Fix space field regression and version change bug

- Restore space field in selectedTimeseries interface (was incorrectly removed)
- Add inline comment explaining why we need space field twice:
  * Top-level space (e.g. 'cdf_cdm') = used for searching/listing in DMS view
  * selectedTimeseries.space (e.g. 'cdm_try') = where actual instance lives for data queries
- Fix version change bug: reset selectedTimeseries when version changes
- Fix search input reset: add dynamic key to AsyncSelect component

muradsater committed:
Fix API error display and improve error handling

- Import and use existing stringifyError utility function
- Fix API error messages showing as [object Object] instead of actual error text
- Now properly displays API errors like 'One or more views do not exist: cdf_cdm:CogniteTimeSeries/v2'
- Add null safety with optional chaining for dropdown clear button errors
- Change error alert styling from red to blue (info level) for less alarming appearance

muradsater committed:
Fix timeseries datasource integration

- Update TimeseriesDatasource to handle CogniteTimeSeries search queries
- Use proper space field from selectedTimeseries for data queries
- Ensure correct request body structure for DMS timeseries data retrieval

muradsater committed:
Update test to use proper DMS types import

- Fix test imports to use DMS types from new location
- Update test structure to match new type organization

muradsater committed:
Fix field spacing and UI consistency

- Move LatestValueCheckbox inside form group for consistent spacing
- Move aggregation fields closer to Latest Value checkbox
- Remove extra spacing between fields for better visual consistency
- Update dashboard configuration to reflect improved spacing

muradsater commented:
## Space Field Regression Fix

**Issue:** I initially removed the  field from the  interface thinking it was duplicate/redundant since we already have a top-level  field. This caused a regression where the request body was malformed and data queries failed.

**Root Cause:** We actually need **two different space fields** for different purposes:

1. **Top-level space** (e.g., ) - Used for searching/listing timeseries in the DMS view
2. **selectedTimeseries.space** (e.g., ) - Where the actual selected instance lives for data queries

**Fix:** 
- Restored the  field in  interface 
- Added inline comment in code explaining why we need both spaces
- Updated request body to use the correct space for data retrieval

**Commits:**
- Space field fix: `c98b843` - Fix space field regression and version change bug
- Code comment: Added inline explanation in `handleTimeseriesSelection`

This ensures the timeseries data queries use the correct space where the instance actually lives, not just the search space.

muradsater commented:
## Space Field Regression - Technical Explanation

**The Problem:**
Initially removed the space field from selectedTimeseries interface thinking it was duplicate since we already have a top-level space field. This caused a regression where data queries failed.

**Why We Need Two Space Fields:**
- **Top-level space** (cdf_cdm) = Used for searching/listing timeseries in DMS view
- **selectedTimeseries.space** (cdm_try) = Where the actual selected instance lives for data queries

**The Fix:**
- Restored space field in selectedTimeseries interface
- Added inline code comment explaining the dual purpose  
- Fixed request body to use correct space for data retrieval

**Related Commits:**
- c98b843: Fix space field regression and version change bug
- Added explanatory comment in handleTimeseriesSelection function

This ensures timeseries data queries use the correct instance space, not just the search space.

muradsater commented on src/types.ts:
```diff
@@ -99,6 +101,18 @@ export interface FlexibleDataModellingQuery {
+    externalId: string;
+    name?: string;
+  };
+  limit?: number;
```
**Space Field Regression Fix**

To answer your question: We need the space field in selectedTimeseries because we actually need **two different spaces** for different purposes:

1. **Top-level space** (e.g., "cdf_cdm") - Used for searching/listing timeseries in the DMS view
2. **selectedTimeseries.space** (e.g., "cdm_try") - Where the actual selected instance lives for data queries

**What happened:**
I initially removed the space field from selectedTimeseries thinking it was duplicate. This caused a **regression** where the request body was malformed and data queries failed.

**The fix:**
- Restored the space field in selectedTimeseries (commit c98b843)
- Added inline code comment explaining why we need both spaces
- Fixed the request body to use the correct space for data retrieval

The top 3 fields (space, version, externalId) are for the DMS view/data model configuration, while the selectedTimeseries.space is for the actual instance location.

muradsater replied:
Updated json structure to be:
```
          "cogniteTimeSeries": {
            "externalId": "CogniteTimeSeries",
            "instanceId": {
              "space": "cdm_try",
              "externalId": "59.9139-10.7522-current.pressure"
            },
            "space": "cdf_cdm",
            "version": "v1"
          }
```
typescript object
```
export interface CogniteTimeSeries {
  space: string;
  version: string;
  externalId: string;
  instanceId?: {
    space: string;
    externalId: string;
  };
}
```

muradsater committed:
Fix lint error: Add missing newline at end of file

- Fix eslint error 'eol-last' in src/types/dms.ts
- Remove trailing space and ensure file ends with newline

muradsater committed:
Fix hardcoded space in timeseries name retrieval

- Replace hardcoded 'cdf_cdm' with dynamic cogniteTimeSeriesSearchQuery.space
- Ensures correct behavior when views from different spaces are selected
- Fixes issue where name property lookup would fail for non-cdf_cdm spaces
- Addresses gemini-code-assist bot feedback on PR #557

muradsater commented:
/gemini help

polomani commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@cognite/cognite-grafana-datasource",
-  "version": "4.3.0",
+  "version": "4.4.0",
```
let's not bump the version yet.
also need to hide this behind the toggle

muradsater committed:
Update query structure to use cogniteTimeSeries and instanceId

- Change cogniteTimeSeriesSearchQuery to cogniteTimeSeries
- Change selectedTimeseries to instanceId
- Update all type definitions and references
- Update weather-station.json dashboard to use new format
- Addresses PR feedback for consistent naming convention

muradsater committed:
Revert to 4.3.0

muradsater committed:
Merge branch 'master' into feature/cognite-timeseries-search-tab

Ratna-Phaneendra-Matukumalli commented on src/components/cogniteTimeSeriesSearchTab.tsx:
```diff
@@ -0,0 +1,287 @@
+import { CommonEditors, LabelEditor } from './commonEditors';
+
+interface CogniteTimeSeriesSearchTabProps extends SelectedProps {
+  connector: any;
```
```suggestion
  connector: Connector;
```

muradsater committed:
Address PR feedback: refactor mock data and improve type safety

- Move mock data to reusable src/__mocks__/dmsTestData.ts for better organization
- Add DMSSearchRequest type annotation to avoid 'as const' assertion
- Change connector type from 'any' to 'Connector' for better type safety
- Clean up test file by using centralized mock data

Addresses feedback from @Ratna-Phaneendra-Matukumalli

muradsater committed:
Fix import path in dmsTestData.ts

- Change import from '../../types/dms' to '../types/dms'
- Resolves TypeScript build error in GitHub CI
- Path was incorrect after moving file to src/__mocks__/

muradsater committed:
Remove 'as const' assertion after adding DMSSearchRequest type

- Remove 'as const' from type: 'view' since DMSSearchRequest type provides proper typing
- TypeScript now properly infers the literal type without manual assertion
- Completes the type safety improvement from PR feedback

Ratna-Phaneendra-Matukumalli reviewed (approved):
LGTM

muradsater committed:
Merge branch 'master' into feature/cognite-timeseries-search-tab
